esmvaltool/diag_scripts/weighting/weighted_temperature_graph.py

```diff
@@ -0,0 +1,204 @@
+
+    plot_shaded(
+        iqr_weighted.time,
+        iqr_weighted.data[:, 0],
+        iqr_weighted.data[:, 1],
+        color=color_weighted,
+        label='Weighted inter-quartile range',
+    )
+
+    for temp in temperature.data:
+        axes.plot(temperature.time,
+                  temp,
+                  color=color_data,
+                  lw=0.5,
+                  alpha=0.5,
+                  zorder=1,
+                  label='Ensemble members')
+
+    # Fix duplicate labels
+    handles, labels = plt.gca().get_legend_handles_labels()
+    by_label = dict(zip(labels, handles))  # dict removes dupes
+    axes.legend(by_label.values(), by_label.keys())
+
+    plt.title('Temperature anomaly relative to 1981-2010')
+    plt.xlabel('Year')
+    plt.ylabel(r'Temperature anomaly $\degree$C')
+
+    filename = get_plot_filename('temperature_anomaly_graph', cfg)
+    plt.savefig(filename, dpi=300, bbox_inches='tight')
+    plt.close()
+
+    caption = 'Temperature anomaly relative to 1981-2010'
+
+    log_provenance(caption, filename, cfg, provenance_info)
+
+    logger.info('Temperature anomaly plot stored as %s', filename)
+
+
+def calculate_percentiles(data: 'xr.DataArray',
+                          percentiles: list,
+                          weights: dict = None):
+    """Calculate (weighted) percentiles.
+
+    Calculate the (weighted) percentiles for the given data.
+
+    Percentiles is a list of values between 0 and 100.
+
+    Weights is a dictionary where the keys represent the model members,
+    and the values the weights.
+    If `weights` is not specified, the non-weighted percentiles are calculated.
+
+    Returns a DataArray with 'percentiles' as the dimension.
+    """
+    if weights:
+        weights = [weights[member] for member in data.model_ensemble.values]
+
+    output = xr.apply_ufunc(weighted_quantile,
+                            data,
+                            input_core_dims=[['model_ensemble']],
+                            output_core_dims=[['percentiles']],
+                            kwargs={
+                                'sample_weight': weights,
+                                'quantiles': percentiles / 100
+                            },
+                            vectorize=True)
+
+    output['percentiles'] = percentiles
+
+    return output
+
+
+def main(cfg):
+    """Plot weighted temperature graph."""
+    input_files = cfg['input_files']
+    filename = cfg['weights']
+    weights_path = Path(input_files[0]) / filename
+    weights = read_weights(weights_path)
+
+    models = read_metadata(cfg, projects=['CMIP5'])['tas']
+    model_data, model_provenance = read_model_data(models)
+
+    percentiles = np.array([25, 75])
+
+    iqr = calculate_percentiles(
+        model_data,
+        percentiles,
+    )
+
+    iqr_weighted = calculate_percentiles(
+        model_data,
+        percentiles,
+        weights=weights,
+    )
+
+    visualize_temperature_graph(
```

Comment:
Usually the plotted data is also saved to a netcdf file, it would be good to do that here too

Reply:
In this case the plotted data doesn't differ much from what is already saved in the preprocessed input data. It doesn't really make sense to copy that file, does it? In the other diagnostic we can (and will) save a few small files for the individual performance and independence scores, just like we already save the weights.

Reply:
Yes, it does make sense if you want to e.g. display the result on a website with an interactive viewer. The preproc dir is usually removed after a successful run, so it's no longer available and also it's not clear that what is in there is useful or interesting to the user of the recipe.